pyuvdata/uvdata/ms.py

```diff
@@ -126,11 +886,715 @@ def find_not_ascii(s):
+        # multi-object support.
+        if self.phase_type == "phased":
+            pass
+        elif self.phase_type == "drift":
+            if force_phase:
+                print(
+                    "The data are in drift mode and do not have a "
+                    "defined phase center. Phasing to zenith of the first "
+                    "timestamp."
+                )
+                phase_time = Time(self.time_array[0], format="jd")
+                self.phase_to_time(phase_time)
+            else:
+                raise ValueError(
+                    "The data are in drift mode. "
+                    "Set force_phase to true to phase the data "
+                    "to zenith of the first timestamp before "
+                    "writing a measurement set file."
+                )
+        else:
+            raise ValueError(
+                "The phasing type of the data is unknown. "
+                "Set the phase_type to drift or phased to "
+                "reflect the phasing status of the data"
+            )
+
+        # Initialize a skelton measurement set
+        ms = self._init_ms_file(filepath)
+
+        # All all the rows we need up front, which will allow us to fill the columns
+        # all in one shot.
+        ms.addrows(self.Nblts * self.Nspws)
+
+        attr_list = ["data_array", "nsample_array", "flag_array"]
+        col_list = ["DATA", "WEIGHT_SPECTRUM", "FLAG"]
+
+        # Some tasks in CASA require a band-representative (band-averaged?) value for
+        # the weights and noise for all channels in each row in the MAIN table, which
+        # we will roughly calculate in temp_weights below.
+        temp_weights = np.zeros((self.Nblts * self.Nspws, self.Npols), dtype=float)
+
+        for attr, col in zip(attr_list, col_list):
+            val_dict = {}
+            temp_array = getattr(self, attr)
+
+            if not self.future_array_shapes:
+                temp_array = np.squeeze(temp_array, axis=1)
+
+            for idx, spw_id in enumerate(self.spw_array):
+                if self.flex_spw:
+                    sel_mask = self.flex_spw_id_array == spw_id
+                else:
+                    sel_mask = np.ones(self.Nfreqs, dtype=bool)
+
+                temp_dict = {
+                    "%016d"
+                    % ((self.Nspws * jdx) + 1 + idx): temp_array[None, jdx, sel_mask, :]
+                    for jdx in np.arange(self.Nblts)
+                }
+                if col == "WEIGHT_SPECTRUM":
+                    temp_weights[idx :: self.Nspws] = np.median(
+                        temp_array[:, sel_mask], axis=1
+                    )
+
+                val_dict.update(temp_dict)
+            ms.putvarcol(col, {key: val_dict[key] for key in sorted(val_dict.keys())})
+
+        # TODO: If/when UVData objects can store visibility noise estimates, update
+        # the code below to capture those.
```

Comment:
Fascinating. I didn't realize there was a place for these in measurement sets. We actually have something like this for HERA.

Reply:
If you can think of it, CASA has an N-dimensional table ready to store said information for you, whether you like it or not ;)